docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,331 @@
+- Kubeadm COULD allow opt-out of kubeadm setting up `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+This is deferred to a future Kubeadm design and release, and for this proposal, we will add fields to KubeadmControlPlane to remove these node groups and bindings post control plane initialisation.
+
+#### Changes to KubeadmControlPlane
+
+```
+spec:
+  disableKubeadmAutoApproveClusterRoleBinding: true
+```
+
+This field on KCP will `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+
+#### Kubelet authenticator flow
+
+The authenticator will be responsible for both updating the kubelet client and serving certificates.
+
+Both of these decision flows will operate inside the authenticator, and execute whenever called by kubelet to provide client authentication.
+
+##### Client authentication flow
+
+![client auth](images/kubelet-authentication/client-authenticator-flow.png)
+
+##### Serving certificate flow
+
+![serving cert flow](images/kubelet-authentication/serving-certificate-flow.png)
+
+### Risks and Mitigations
+
+There may be additional security risks being introduced in this design. In order to mitigate this, this proposal will be taken to SIG Security and SIG Auth for review **before the beta graduation**.
+
+## Alternatives
+
+#### Implement within the cloud providers instead of Cluster API
+Given that there is an existent implementation in cloud-provider-gcp, this could be extended to all of the cloud providers. However, there are some advantages to making Cluster API responsible for kubelet registration in that no changes to the assumptions around connectivity between management and workload clusters are required, neither does the signer need to be included as a static pod during control plane instantiation.
+
+#### Implement as authentication webhook, as per aws-iam-authenticator (Amazon EKS)
+If attestation was implemented as an authentication webhook, it would be in the critical path for all token-based authentication against the API server. It would also additionally be needed to be set up at workload cluster instantiation via a static pod and API server start up.
+
+
+## Upgrade Strategy
+
+for the upgrade strategy please refer to the Cluster API node bootstrapper proposal.
+
+## Additional Details
+
+### Test Plan [optional]
+
+- E2E tests to be added to AWS, vSphere and Azure providers as each provider implements the signer
+- E2E tests to be added for the insecure signers for use with CAPD.
+Upgrade tests from latest minor release to latest main branch of Kubernetes
+
+
+### Graduation Criteria [optional]
+
+#### Graduation to beta
+- E2E tests testing upgrades to latest main branch of Kubernetes are required such that Cluster API can make appropriate changes to node registration if kubelet or kubeadm behaviour changes.
+
+- Security review by SIG Auth and SIG Security
+#### Graduation to GA
+- Beta for 2 releases
```

Comment:
What is this assumption based on?

Reply:
This was thin-air boilerplate from the old google doc. Open to suggestions here.

Reply:
I think it will be hard to specify *now* because there seem to be a lot of unknowns about future releases, stability guarantees etc.